src/main/java/org/broadinstitute/dsde/consent/ontology/datause/services/TextTranslationServiceImpl.java

```diff
@@ -89,13 +98,70 @@ public String translateDataset(String dataUseString) throws IOException {
         return translate(dataUse, TranslateFor.DATASET);
     }
 
+    @Override
+    public HashMap<String, Recommendation> translateParagraph(String paragraph) throws Exception {
+        return paragraph(paragraph);
+    }
+
     @Override
     public String translatePurpose(String dataUseString) throws IOException {
         ObjectMapper mapper = new ObjectMapper();
         DataUse dataUse = mapper.readValue(dataUseString, DataUse.class);
         return translate(dataUse, TranslateFor.PURPOSE);
     }
 
+    private HashMap<String, Recommendation> paragraph(final String paragraph) throws Exception {
+      HashMap<String, Recommendation> recommendations = new HashMap<>();
+
+      List<TermItem> terms = loadTermsFromGoogleStorage();
+
+      if (terms.isEmpty()) {
+        throw new Exception("Unable to process search terms");
+      }
+
+      for (TermItem term : terms) {
+        final String title = term.getTitle();
+        final String category = term.getCategory();
+        final String url = term.getUrl();
+        final String[] keywords = term.getKeywords();
+
+        for (String keyword : keywords) {
+          final boolean foundMatch = searchForKeyword(keyword, paragraph);
+
+          if (foundMatch) {
+            Recommendation recommendation = new Recommendation(
+                title,
+                category
+            );
+            if (!recommendations.containsKey(url)) {
+              recommendations.put(url, recommendation);
+            }
```

Comment:
Since `if (!containsKey) { put(...); }` is a common form there's a java method for it, `putIfAbsent`. You can write this code as
```suggestion
            recommendations.putIfAbsent(url, new Recommendation(title, category));
```

If you want to avoid making a new object in the case where the map already contains a key, you can use `computeIfAbsent`
```suggestion
            recommendations.computeIfAbsent(url, key -> new Recommendation(title, category));
```